tr2/vars: remove global object related vars

lahm86 opened the pull request:
#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This eliminates some global object related variables in favour of keeping scope tighter to their linked modules. The final level counter object has also been renamed, and its handling is now managed fully in the same module. The shower shotgun needs a reference to one of the variables still - I considered moving this object's control into the same module as it's so tied together but I've left that for now; I think it's OK to have an API for this bit.

Testing:
- Detonating all mines in Venice e.g. if one blows up, the rest should follow suit
- Home Sweet Home should behave as normal, so starting the level from fresh, from a save before the boss, a save with the boss alive, and another with him defeated

lahm86 committed:
tr2/vars: move mine detonation flag to module scope

This moves the global flag to detonate all mines into the mines module.
It is reset during object setup.

lahm86 committed:
tr2/objects/combat_end: rename final level counter

This loosens the tie somewhat of the Home Sweet Home kill-to-complete
logic by renaming the final level counter to "combat end".

lahm86 committed:
tr2/objects/combat_end: move initialisation to module scope

This eliminates the direct setup function for combat end and moves it
to regular item initialisation. Object setup is responsible for
resetting the relevant variables.

lahm86 committed:
tr2/objects/combat_end: eliminate global boss item array

This replaces the global boss item array with logic to find the best
boss based on the level array at the required time.

lahm86 committed:
tr2/objects/combat_end: move global vars to module scope

This removes the global variables related to the final level and
creates a function for the shower cutscene shotgun control to check if
the end timer is complete.

rr- commented on src/tr2/game/objects/general/combat_end.c:
```diff
@@ -10,12 +10,26 @@
+    O_WORKER_3,
+    NO_OBJECT,
+    // clang-format on
+};
```
I'm thinking, maybe this could be enhanced to test for any enemy type except `M_BOSS_TYPE`?

lahm86 replied:
Thanks, that's much neater. I've shared a quick test level for verifying (`hsh-any-enemies`) - @aredfan, if you wouldn't mind giving it a quick check over.

I think the boss type could be moved to either scripting, or potentially when we're adding killable count stats, we can do something here (i.e. those without triggers).

lahm86 committed:
tr2/objects/combat_end: accept all regular enemy types

This allows the combat end logic to work with any regular enemies
rather than being fixed to the dog and two goon types. The boss type
currently remains hard-coded.

rr- reviewed (commented):
The proposed patch will count allies – I think we'll need g_EnemeyObjects.

lahm86 commented:
Good catch, thanks.
